processor/PlanExecutor.go

```
@@ -18,43 +22,102 @@ func merge(source map[string]interface{}, extra map[string]interface{}) map[stri
-		Bar:          bar,
-		Publisher:    telegraph.NewLinkedPublisher(),
-		PlanContext:  core.ExtractionResult{},
-		JobContexts:  map[int]core.ExtractionResult{},
-		StepContexts: map[int]map[int]core.ExtractionResult{},
-		mutex:        &sync.Mutex{},
+func (instance *PlanExecutionContext) execute(cancellation chan struct{}) {
+	instance.workerExecuteJobs(instance.Plan.Jobs, cancellation)
+}
+
+func (instance *PlanExecutionContext) workerExecuteJobs(jobs []core.Job, cancellation chan struct{}) {
+	var jobStream JobStream
+	jobStream = CreateJobSequentialStream(jobs)
+
+	if instance.Config.Random {
+		jobStream = CreateJobRandomStream(jobs)
+	}
+
+	if instance.Config.Iterations > 0 {
+		revolvingStream := CreateJobRevolvingStream(jobStream)
+		iterationStream := CreateJobIterationStream(*revolvingStream, len(jobs), instance.Config.Iterations)
+		jobStream = iterationStream
+	}
+
+	if instance.Config.Duration > time.Duration(0) {
+		jobStream = CreateJobDurationStream(jobStream, instance.Config.Duration)
+		ticker := time.NewTicker(time.Millisecond * 10)
+		//		go func() {
+		//			for _ = range ticker.C {
+		if int64(time.Since(instance.start).Seconds())%2 == 0 {
+			instance.progress <- jobStream.Progress()
+		}
+		//			}
+		//		}()
+		time.AfterFunc(instance.Config.Duration, func() {
+			ticker.Stop()
+			_ = instance.Bar.Set(100)
+		})
+	}
+
+	for jobStream.HasNext() {
+		job := jobStream.Next()
+		_ = instance.Bar.Set(jobStream.Progress())
+		//before Job
+		for _, action := range job.Before {
+			_ = action.Execute(nil, cancellation)
+		}
+		instance.workerExecuteJob(job, cancellation)
+		//after Job
+		for _, action := range job.After {
+			_ = action.Execute(nil, cancellation)
+		}
 	}
 }
 
-func (instance *PlanExecutor) executeStep(step core.Step, cancellation chan struct{}) core.ExecutionResult {
-	start := time.Now()
+func (instance *PlanExecutionContext) workerExecuteJob(talula core.Job, cancellation chan struct{}) {
+	defer func() { //catch or finally
```

Could we push this func onto the `errormanager` called something like `Handle`, so we can make more liberal use of `defer errormanager.Handle()` in other places?

---

Not sure about this one.  TBH I shouldn't be swallowing the error and a generic Handle function for errors is covered with `errormanager.Check(...)`.  I am actually not sure why I am catching the error in this method and not letting it propagate.  Going to remove this code block for now.